### ISSUE TITLE
BUG: pagination in GridField does not work if DataObject default_sort field values are not unique

### ISSUE THREAD

**Issue by sunnysideup** [silverstripe/silverstripe-framework#7249]
# SUMMARY:

Pagination does not work well when the data is provided by MySQL 5.6 / 5.7 (php 7.1) where the data-set is sorted by a field (or fields) that are not enforced to be unique on a database level.  The reason for this is that MySQL may return datasets in a random order (beyond what it has been asked to sort by). 

You can possibly replicate this bug by yourself by using the tools listed below:
 * https://github.com/sunnysideup/silverstripe-sort-test/: see the bug in action
 * https://github.com/sunnysideup/silverstripe-dataintegritytests/blob/master/code/CheckForMysqlPaginationIssuesBuildTask.php: analyse your own project

This means, for example, that:
* batch processing may not work correctly (you will process some records twice and other ones not at all)
* grid field pagination is broken (some items listed twice or more, on separate pages, others not at all)
* pagination on the front-end is broken (some items listed twice or more, on separate pages, others not at all)

There are still a lot of questions about this bug, but I have been able to consistently demonstrate this error on several machines with clean installs of 3.6.1

# Details
the issue: when the default_sort value for a dataobject contains non-unique values for one or more rows then the GridField ends up NEVER showing some records while showing other records more than once:

How to replicate:

1. install SS 3.6.1 using standard installer

2. add the following dataobject (`MyDataObject`):

```php

<?php


class MyDataObject extends DataObject
{

    private static $db = array(
        'Title' => 'Varchar(20)',
        'SameSame' => 'Varchar(20)'
    );


    private static $default_sort = array(
        'SameSame' => 'ASC'
    );

    function requireDefaultRecords()
    {
        DB::query('DELETE FROM MyDataObject');
        for($i = 1; $i < 47; $i++) {
            $filter = array('Title' => "item #".$i);
            if(! MyDataObject::get()->filter($filter)->first()) {
                $obj = MyDataObject::create($filter);
                $obj->SameSame = 'same';
                $obj->write();
            }
        }
    }

}
```

then add modeladmin (`MyModelAdmin.php`):

```php

<?php

class MyModelAdminTest extends ModelAdmin {

    private static $managed_models = array('MyDataObject');

    private static $url_segment = 'test';

    private static $menu_title = 'Test';
}
```

You get the following results:
![image](https://user-images.githubusercontent.com/167154/29008421-bd164ac2-7b6a-11e7-8689-412c2d973cf6.png)

![image](https://user-images.githubusercontent.com/167154/29008426-c8201c2c-7b6a-11e7-9f39-6776f680b72d.png)

For my client, I was sorting a GridField by EndDate and StartDate. Some of the items has the same EndDate and StartDate and thus were missing.  Making it impossible for them to be found by the client.

**Comment by flamerohr** [comment]
Have you tried adding `ID` as the last item in the list?
This does look like a genuine bug, but perhaps adding ID would be a workaround for it

**Comment by sunnysideup** [comment]
@flamerohr  - you of course that is a work around, but this is a serious bug - because anytime you do not sort by a field that mysql enforces to be unique, i.e. in basically almost all cases (e.g. sorting by Title, by Created, by LastEdited, by Date, and so on....  you will potentially end up with missing records in the GridField.

**Comment by flamerohr** [comment]
unfortunately, this is a potential problem in any given relational database, many queries do not explicitly provide a sort (or not a sort with at least a unique column) and has the "implied" sorting by ID, which is what happened here.

I've marked as critical because of data disappearing, but I encourage always adding explicit `ID` sort to the end of any sort calls.

**Comment by unclecheese** [comment]
I can't replicate this on `3.6.1`. Can you test it on a plain install?

**Comment by sunnysideup** [comment]
I run this on a plain 3.6.1

**Comment by unclecheese** [comment]
I wonder if the auto increment on your ID is isn't working properly. Can you try deleting the entire DB and starting over? I'm curious if you're getting IDs that are non sequential with your Item numbers due to all the deletions. Try adding `ID` as a `summary_field` to help debug.

**Comment by unclecheese** [comment]
![screenshot 2017-08-08 10 53 58](https://user-images.githubusercontent.com/654636/29049043-ed00fb98-7c27-11e7-916d-4061905d250f.png)

**Comment by sunnysideup** [comment]
No problems with IDs


Title | ID
-- | --
item #1 | 334
item #2 | 335
item #3 | 336
item #4 | 337
item #5 | 338
item #6 | 339
item #7 | 340
item #8 | 341
item #9 | 342
item #10 | 343
item #11 | 344
item #12 | 345
item #13 | 346
item #14 | 347
item #15 | 348
item #16 | 349
item #17 | 350
item #18 | 351
item #19 | 352
item #20 | 353
item #21 | 354
item #22 | 355
item #23 | 356
item #24 | 357
item #25 | 358
item #26 | 359
item #27 | 360
item #28 | 361
item #29 | 362
item #30 | 363
item #31 | 364
item #32 | 365
item #33 | 366
item #34 | 367
item #35 | 368
item #36 | 369
item #37 | 370
item #38 | 371
item #39 | 372
item #40 | 373
item #41 | 374
item #42 | 375
item #43 | 376
item #44 | 377
item #45 | 378
item #46 | 379

**Comment by unclecheese** [comment]
What kind of output do you get when you just run `SELECT Title FROM MyDataObject ORDER BY SameSame ASC` in the database?

**Comment by sunnysideup** [comment]
I run the following code:

```php
    public function MyDataObjectRaw()
    {
        $this->html .= '<br />-------------------------------------';
        $rows = DB::query('SELECT Title, ID FROM MyDataObject ORDER BY SameSame ASC;');
        foreach($rows as $row) {
            $this->html .= '<li>'.$row['Title'].' = '.$row['ID'];
        }
        $this->html .= '<br />-------------------------------------';
        $rows = DB::query('SELECT Title, ID FROM MyDataObject ORDER BY SameSame ASC LIMIT 0, 10;');
        foreach($rows as $row) {
            $this->html .= '<li>'.$row['Title'].' = '.$row['ID'];
        }
        $this->html .= '<br />-------------------------------------';
        $rows = DB::query('SELECT Title, ID FROM MyDataObject ORDER BY SameSame ASC LIMIT 10, 10;');
        foreach($rows as $row) {
            $this->html .= '<li>'.$row['Title'].' = '.$row['ID'];
        }
        $this->html .= '<br />-------------------------------------';
        $rows = DB::query('SELECT Title, ID FROM MyDataObject ORDER BY SameSame ASC LIMIT 20, 10;');
        foreach($rows as $row) {
            $this->html .= '<li>'.$row['Title'].' = '.$row['ID'];
        }
    }

```

and the output that I get is:



-------------------------------------
item #36 = 369
item #25 = 358
item #26 = 359
item #27 = 360
item #28 = 361
item #29 = 362
item #30 = 363
item #31 = 364
item #32 = 365
item #33 = 366
item #34 = 367
item #35 = 368
item #24 = 357
item #37 = 370
item #38 = 371
item #39 = 372
item #40 = 373
item #41 = 374
item #42 = 375
item #43 = 376
item #44 = 377
item #45 = 378
item #46 = 379
item #13 = 346
item #2 = 335
item #3 = 336
item #4 = 337
item #5 = 338
item #6 = 339
item #7 = 340
item #8 = 341
item #9 = 342
item #10 = 343
item #11 = 344
item #12 = 345
item #1 = 334
item #14 = 347
item #15 = 348
item #16 = 349
item #17 = 350
item #18 = 351
item #19 = 352
item #20 = 353
item #21 = 354
item #22 = 355
item #23 = 356
-------------------------------------
item #1 = 334
item #2 = 335
item #3 = 336
item #4 = 337
item #5 = 338
item #6 = 339
item #7 = 340
item #8 = 341
item #9 = 342
item #10 = 343
-------------------------------------
item #12 = 345
item #1 = 334
item #10 = 343
item #9 = 342
item #8 = 341
item #7 = 340
item #6 = 339
item #5 = 338
item #4 = 337
item #3 = 336
-------------------------------------
item #12 = 345
item #11 = 344
item #10 = 343
item #9 = 342
item #8 = 341
item #7 = 340
item #6 = 339
item #5 = 338
item #4 = 337
item #3 = 336

**Comment by unclecheese** [comment]
I'm not convinced this is a SilverStripe issue. The way that MySQL resolves an ambiguous sort is highly unpredictable. See (https://dba.stackexchange.com/questions/6051/what-is-the-default-order-of-records-for-a-select-statement-in-mysql).

> MySQL sorts the rows any way it wants to, without any guarantee of consistency.

That I can't replicate it should be an indication that we're dealing with some kind of unknown. It could be as simple as differing storage engines, MySQL versions, or both. In short, don't use ambiguous sorts.

**Comment by sunnysideup** [comment]
Hi Aaron,

> In short, don't use ambiguous sorts.

That is true in theory, but I dont think that is true practice, most of the time you are sorting by an ambiguous sort: e.g. we tend to sort by Sort / SortNumber (which are usually not database enforced uique), Title, Name, Calendar Start Date, Created, LastEdited, etc... etc....  Anytime they are not unique, you will end up with this problem. 

So, I reckon we should fix this, for the following reasons:

(1)
I am sure that most `default_sort` entries, such as SiteTree.php do not sort by a database enforced unique key. If any of them are displayed in a paginated list (e.g. a gridfield, search results, filtered list, etc... etc... ) there is a real chance that some will show up twice on different pages and that other items will never show. 

(2) There is an easy fix - add an ORDER BY BaseTable.ID to the end of all SQL Sort statements

(3) The danger with this bug is that no one will notice until it has real life consequences.  There is no error, there lists seem to operate fine, etc... 

(4) A perfect developer should remember to add the Sort By ID to the end of the sort statement - but if (a) you always need to add this (b) few developers actually do this in practice - then this is another reason for adding it.

**Comment by sunnysideup** [comment]
I could *not* replicate this behaviour on my lappy:

+-------------------------+-------------------------+
| Variable_name           | Value                   |
+-------------------------+-------------------------+
| innodb_version          | 5.7.19                  |
| protocol_version        | 10                      |
| slave_type_conversions  |                         |
| tls_version             | TLSv1,TLSv1.1           |
| version                 | 5.7.19-0ubuntu0.16.04.1 |
| version_comment         | (Ubuntu)                |
| version_compile_machine | x86_64                  |
| version_compile_os      | Linux                   |
+-------------------------+-------------------------+

I will check mysql version at work.

**Comment by sunnysideup** [comment]
Here is another test I ran:

Page:

```php


    function requireDefaultRecords()
    {
        for($i = 1; $i < 400; $i++) {
            DB::alteration_message('Creating Pages: '.$i);
            $filter = ['Title' => 'Page No. '.$i];
            $page = Page::get()->filter($filter)->first();
            if(! $page) {
                $page = Page::create($filter);
            }
            if($i > 10) {
                $parentPage = Page::get()->filter(array('ID:LessThan' => 10))->sort('RAND()')->first();
                if($parentPage) {
                    $page->ParentID = $parentPage->ID;
                }
            } else {
                $page->ParentID = 0;
            }
            $page->writeToStage('Stage');
            $page->publish('Stage', 'Live');
        }
    }
```

Page_Controller:
```php


    public function UnpaginatedPages()
    {
        return Page::get();
    }

    public function PaginatedPages()
    {
        $list = Page::get();

        $list = new PaginatedList($list, $this->getRequest());
        $list->setPageLength(395);
        return $list;
    }
```

Page.ss

```html
    <% loop PaginatedPages %>
        <li>$Title</li>
    <% end_loop %>

    <% if $PaginatedPages.MoreThanOnePage %>
        <% loop $PaginatedPages.Pages %>
            <% if $CurrentBool %>
                $PageNum
            <% else %>
                <% if $Link %>
                    <a href="$Link">$PageNum</a>
                <% else %>
                    ...
                <% end_if %>
            <% end_if %>
            <% end_loop %>
    <% end_if %>
```

**Comment by unclecheese** [comment]
> (2) There is an easy fix - add an ORDER BY BaseTable.ID to the end of all SQL Sort statements

I really think this is dangerous scope creep for the ORM, and it sets a really bad precedent. The ORM is an abstraction layer between you and SQL. It's primary purpose is to make querying more succinct, more secure, and more declarative. It is not the ORM's job to fix your bugs. It should be very possible to write bad queries with the ORM, and sorting by something you know not to have a singular value is one of those areas. 

> (a) you always need to add this no matter what

Personally, I've never had to do this. Keep in mind, you're testing a very extreme case, where all the values of a column are equal, and in any real-world scenario, it would be apparent to the developer that he was sorting on a column with uniform values. If, by contrast, I sort by `Title ASC`, and 2/100 records have the same title, I'm not sure I've ever noticed any consequences for the logic MySQL uses to resolve that conflict. If I did, I'd add a second column.

**Comment by tractorcow** [comment]
The best practice, in these scenarios, is to encourage developers to implement composite default_sort values for their models, if they are aware that their sort column is not unique. I concur with @unclecheese on this.

E.g.

```php
private static $default_sort = '"BaseTable"."Sort" ASC, "BaseTable"."ID" ASC';
```

**Comment by flamerohr** [comment]
I agree, documentation to encourage this would be the best outcome for this...

I don't believe enforcing an `ID` sort at the end of all queries which uses a sort is a good idea.

The `SiteTree.default_sort` also not having `ID` doesn't mean it's best practice :) in fact if anyone created a pull request to add `ID` in for `SiteTree`, I think it'd be a welcomed change.

**Comment by sunnysideup** [comment]
>The best practice, in these scenarios, is to encourage developers to implement composite default_sort values for their model

@tractorcow 

What about `SiteTree.Sort` and  Blog:

https://github.com/silverstripe/silverstripe-blog/blob/master/src/Model/BlogPost.php#L134-L139

Both models are bound to have non-unique sort values.

**Comment by tractorcow** [comment]
We could add ID to those models exclusively, but that's still preferable to baking it into datalist for all classes globally.

**Comment by sunnysideup** [comment]
Check out my scenario:

On an e-commerce website we offer discount vouchers. We had set the default_sort to `EndDate DESC, StartDate DESC` as that made sense at the time. In the modelAdmin that works great. 

Then the client adds a whole bunch of discount vouchers with the same start and end dates.

When he is trying to delete them, there are around 8 / 46 that  are not showing up in the ModelAdmin (and another 8 that show up twice) and calls me. I immediately add the `ID` value and that sorted it.  However, as you can see, BlogPost and Page have the same issue and this is obvious not ideal. 

Another example
---------------------

Member List for Club, Ms Smith can not find herself in list, calls admin, admin finds her on page three of list, sends her a screenshot and a link to page three. Ms Smith can not find herself on that page (there are a few Smith people, list is sorted by last name).

**Comment by kinglozzer** [comment]
`SiteTree.Sort` isn’t supposed to be unique, is it? The sort is relative to the parent page

**Comment by sunnysideup** [comment]
@kinglozzer no, `SiteTree.Sort` is not supposed to be unique, but what I found is that when you sort by a non-unique field (e.g. SiteTree.Sort) then pagination (including GridField pagination) can do pretty crazy things as Mysql is rather unpredictable in its sort order when there are non-unique values.

**Comment by sunnysideup** [comment]
Here is my code as a repository:

https://github.com/sunnysideup/silverstripe-sort-test

I did some further testing.  

It does not happen with pages, it only happens with DataObjects.

**Comment by unclecheese** [comment]
I think this is really getting into a philosophical discussion on what is the role of framework/ORM. That seems to be where we're divided. My thoughts:

If you assign a singular sort clause to a DataObject for a column that is not entirely unique, and it produces unexpected results,  **that's on you** as the developer who wrote that code. To expect the framework to jump in and save you is undue, for the same reason the template parser doesn't close your `</p>` tags.

If we were to implicitly put `ID ASC` in the sort of every `DataList`, we would be punishing every developer who went about it the right way, and unique columns, or composite clauses. There's a performance impact for having multiple sort columns, and it's nothing I would want to impose on every query without a full opt-in from the user. If I'm sorting on something almost guaranteed to be unique, like `Created`, I don't need or want `ID` in the mix as well, and if I have to write special code to tell the framework not to do that, that creates a really suboptimal developer experience.

**Comment by sunnysideup** [comment]
A Framework is there to make its users happy. 
- add sort by ID to GridField paginated lists?
- add sort by ID to Paginated lists (optional)?

Also - `Created` is often not unique (e.g. import of data - import through API - high volume sales).

**Comment by unclecheese** [comment]
Both `GridField` and `PaginatedList` are ORM agnostic. They're designed to work only with instances of `SS_List`, so injecting concerns over `ID` into either of them is inappropriate.

The fact that many users would be punished by this safety feature is really a non-starter for me. I don't think it aligns well with the core values of the product.

**Comment by sunnysideup** [comment]
I tried sorting by `Created` - same issue (most objects are created within one second).

**Comment by unclecheese** [comment]
That's not the point, though. In a real-world scenario where you were using data that was all generated programatically in a matter of milliseconds, you would intuit the `Created` column to be unreliable for sort, and use a different column, or a composite sort clause. What you can demonstrate in test code should not influence the decisions we make that affect the developer experience.

Developers are responsible for understanding the idiosyncrasies of their data, and to install guardrails in the framework to protect them from writing bad code sets a terribly dangerous precedent. If it is acceptable for the ORM to audit the efficacy of your sort clause (at the expense of performance), it logically follows that the framework is permitted to make all kinds of other overreach in the name of developer hand-holding, like closing `<p>` tags, preventing redirect loops, removing unnecessary queries, etc.

**Comment by sunnysideup** [comment]
:-) 

Thank you for all your attention to detail in this bug report.

These are my assumptions right now:

- the problem exists
- it may negatively affect our users in a significant way 
- most Silverstripe code out there does not follow best practice in relation to this issue (i.e. most code does not add a Sort By Unique Field to pagination / gridfields).

Reply from SS core team is: fix your own code, it is not good to punish best practice by adding a `hack` for sloppy code (always sorting by ID). 

I think we need to test the assumptions a bit more first.  I have set up a small repo other developers can see the bug in action: https://github.com/sunnysideup/silverstripe-sort-test/. 

Thank you again for looking into all of this.

**Comment by unclecheese** [comment]
Have added an actionable issue to the backlog.

**Comment by unclecheese** [comment]
Closing as

**Comment by sunnysideup** [comment]
I just realised then when you sort by one the headers in a gridfield, you end up reproducing the same bug.

I am now sorting my dataobjects by ID:

![image](https://user-images.githubusercontent.com/167154/29162080-7c87fc0e-7e0c-11e7-892f-da072b104dcf.png)


But when I click on SameSame, you end up having some items that never show up and other items that show up twice on two pages.


I still have a nagging feeling that the consequences of this bug are underestimated.

![image](https://user-images.githubusercontent.com/167154/29162152-a8fec01a-7e0c-11e7-9b0f-ebf2a82c16ba.png)


page 2:

![image](https://user-images.githubusercontent.com/167154/29162168-ba114a62-7e0c-11e7-8622-8f0b6b242a69.png)

**Comment by patricknelson** [comment]
I get that this is an issue when you have duplicate field values. For me, the best practice when it was known that you could have duplicate values and it was important to have _consistent and predictable_ ordering (particularly in the case of pagination), I was always certain to include a secondary sort field to reduce ambiguity at the database level.  That said, I just wanted to point out:

> ... but sort by ID ASC/DESC should always be added as the last sort criteria, no matter what. 

I would have to disagree with this, since it I think it potentially adds an extra level of magic ✨ which is the last thing you want in an already complicated ORM. What happens when the ORM incorrectly parses/interprets the user's intent in their `$default_sort` property? How do we know they want older or newer records first (when there are dupes)?  What happens if you need to use `FIELD(...)` or `RAND(...)` functions? I think it adds more issues than it solves, since:

1. Adhere to best practice; when sorting order must be consistent, ensure you resolve ambiguity on your own.
2. KISS principle: Let the database do the heavy lifting and reduce complexity as much as possible (increases transparency).

**Comment by sunnysideup** [comment]
@patricknelson, I agree with your theory, but in practice this is not always so easy or practicable.  My thinking right now is that we should add a Sort by ID ONLY when you limit a set, but I have not really found a solution yet.

Have you considered that, for example, by default the GridField allows the user to sort by all Summary Fields that are database fields? Thus, when you sort by a non-unique value in the GridField this bug will raise its ugly head, even though your default_sort may have included the ID field.

**Comment by patricknelson** [comment]
That's `GridField`, however, which allows you to select sort columns. I was replying to:

> ... but sort by ID ASC/DESC should always be added as the last sort criteria, no matter what.

... and then clarified that this shouldn't happen at the ORM (or framework) level when sorting in general. When you're in a `GridField`, I think it might be fine to enable that feature (possibly by default) and then maybe allow disabling of it via some extra setting on your `GridField` instance, if it gets in the way.

**Comment by sunnysideup** [comment]
Hi Patrick,

Thank you for our reply. I think the first thing is really understand the issue, when does it happen, where does it happen. I also jumped to the "solution" but I think the key is to understand the issue first. It does not seem to happen with all DataObjects, so I want to know, for example why that is.

It seems to me when you add DataObjects programatically (i.e. many at the same time) then it is more likely to happen.  Also, I could not replicate it on Pages, why is that?

**Comment by unclecheese** [comment]
We've had a bit of a chat about this internally, and there's some consensus that a halfway measure to mitigate this is appropriate. 

First, the condition of MySQL returning a non deterministic result set when ambiguous sort columns are used is absolutely a feature, not a bug. From the [MySQL docs](https://dev.mysql.com/doc/refman/5.7/en/limit-optimization.html):

> If multiple rows have identical values in the ORDER BY columns, the server is free to return those rows in any order, and may do so differently depending on the overall execution plan. In other words, the sort order of those rows is nondeterministic with respect to the nonordered columns. One factor that affects the execution plan is LIMIT, so an ORDER BY query with and without LIMIT may return rows in different orders.

While MySQL is providing the "expected result" (albeit non deterministic) in this case, it's fair to say DataList is not. As a implementor of the Sortable interface, DataList is expected to sort deterministically. Further, it is the job of the ORM to obscure the "how" from the user, and create the list in a consistent, uniform way, whether you're using MySQL, Postgres, etc, or just plain ArrayData. 

All that said, to force ID into the sort is still too much magic, but ensuring default_sort is added seems like a reasonable halfway point. The problem is, with the direction being included as part of the default sort, it makes sort direction toggling (I.e. GridField column headers) unreliable. Therefore, a more sensible setting might be something more like default_sort_column, with null being allowed to disable forced deterministic sorts. 

Thoughts?

**Comment by sunnysideup** [comment]
Good thoughts ;-) I agree wit hall of that.

Here are a couple of thoughts I had:

* limit without reliable order of records is useless (i.e. `limit 10, 10` may return exactly the same as `limit 0, 10`). 
* database enforced unique fields are not a good option in silverstripe: https://github.com/silverstripe/silverstripe-framework/issues/6819.

Putting these two together ... whenever you use limit in an ORM call, you will have to `sort by ID` as the last sort fragment (e.g. turn `ORDER BY "StartDate" DESC` into `ORDER BY "StartDate" DESC, "ID" ASC`).

So, what I would recommend is: whenever we limit a recordset in an ORM call, we add sort by ID as the last sort phrase (with the ability to turn it off) by default.  This is how I would do it:

When putting together the final SQL for a database query - IF:
* limit is included
* there is no sort by a database enforced unique field
* if the datbase is MySQL (and others???)
* then add `sort by ID` 
* unless the user turns it off

Research questions:

* why does this random sorting no occur on Pages - AFAIK?
* does it also occur on other Databases (apart from MySQL)?
* what is the time difference between `SELECT * FROM Table` AND `SELECT * FROM Table ORDER BY ID` - for a table with a few records and a table with a ton of records...
* is 4.0 also affected?
* are there faster ways to ensure a reliable sort order other than adding ORDER BY ID
* how we do we ensure that there is no ambiguous ID added (i.e. the baseclass / table is always added).

**Comment by unclecheese** [comment]
So the answer to most of your questions/concerns is that it's not a bug or oddity with MySQL or Postgres or pages or DataObjects. This is something that's encoded in the SQL standard. The server is free to return the rows in any order it wants when identical values are encountered. That means, potentially, all database platforms are affected, and the way that this issue may manifest itself in various contexts, be it with the application of certain clauses, implementation of / version of storage engines, or just plain phases of the moon, is irrelevant. The point is that the SQL standard in this case tells us the result is nondeterministic, and we have an API that promises to be deterministic, so we need to reconcile that conflict somehow.

I think if we were to add a `default_sort_column` setting, it would have to be applied to all queries -- not just those that use `LIMIT`. If you had page of results that was all-inclusive, with ambiguous sort, that URL now becomes non-cacheable, as the results could in theory be returned differently on subsequent page loads. The goal should be to make `DataList` deterministic, not to negotiate the random behaviour of your database platform.

**Comment by flamerohr** [comment]
are we able to infer `default_sort_column` from getting the `PrimaryKey` type fields in `DataObjectSchema::databaseFields()`?

Probably a bit too much magic for my taste, but it's an idea to bounce around.

**Comment by patricknelson** [comment]
Since I think this is just in reference to `DataObject`s (where the `default_sort_column` setting will exist), the `ID` column should also  always exist. Is it possible to change change a `DataObject''s primary key? I don't think so, but if it is, then maybe that'd be a concern.

**Comment by sunnysideup** [comment]
How does it work in PHPMyAdmin?

I think we need to do more research. I found that only some tables return with duplicates between limited segments of the table, while other tables do not (e.g. SiteTree).

**Comment by tractorcow** [comment]
A challenge I've considered with a default_sort_column is that there needs to be a way to tell a datalist to reverse it in the case where a non-default sort column (E.g. Title) is reversed, otherwise groups of rows with the duplicate Title column would not reverse. That makes the solution not able to be applied so transparently.

**Comment by sunnysideup** [comment]
*IMPORTANT* 

A:
```sql
SELECT * FROM `SiteTree` LIMIT 20, 10 
```

B: 
```sql
SELECT * FROM `SiteTree` LIMIT 20, 10 ORDER BY ID
```


C: 
```sql
SELECT * FROM `SiteTree` LIMIT 20, 10 ORDER BY `sort`
```

NOT sorting and sorting by ID are equally fast, but sorting by Sort is significantly slower ... see: https://github.com/silverstripe/silverstripe-framework/issues/7272

This evidence shows that sort by ID vs no sort key has no performance impact and sorting by ID is faster than sorting by the default_sort.

**Comment by patricknelson** [comment]
Yeah it's not just in sorting in `ASC`/`DESC` but also if/when you're using functions to sort (e.g. `FIELD(...)` or `IF(...)` and maybe some others, mentioned his above). I don't know what the solution to this would be since this is known behavior, generally speaking. You could always suggest people include a secondary sort column in the meantime (separate issue, update doc's). Depending on how users apply sort, it may be possible to infer the direction if you apply ID as the second column. Often time it's just a second parameter passed to `->sort('Sort', 'DESC')`. It's when you get into custom sort queries that his coildnt be parsed reliably (you could try) but I think at that level the dev should have more control with less magic. 

Ask: Yeah @sunnysideup, typically sorting by indexed fields is a bit faster than unindexed.

**Comment by chillu** [comment]
OK, there's about twenty screenpages of discussion here ;) If you care strongly about resolution, can somebody please summarise the state of the discussion with options going forward?

**Comment by unclecheese** [comment]
Summary:

* When an ambiguous sort column is used, the database is free to return the rows in any order it wants, making the operation non-deterministic.
* This makes pagination unpredictable (e.g. repeated entries across pages)
* As an implementor of `Sortable`, `DataList` _should_ sort deterministically.
* Therefore, we should patch it
* Forcing `ID ASC` into the query is too much magic. Something like `default_sort_column` would be more appropriate
* The bug has existed at least since SS2, and no one raised it until now, so `impact/low` is fitting.

**Comment by patricknelson** [comment]
> Forcing ID ASC into the query is too much magic. Something like default_sort_column would be more appropriate

I think we need to disambiguate the use (or the nomenclature) of `default_sort_column` since we already have `default_sort` which can define both column and direction. As such, since we're addressing a specific problem (deterministic sorting) I think we should take a specific approach that makes it a bit more transparent, **both** in nomenclature and in implementation.

So, I'd suggest we do the following:

- Utilize `ID` as our secondary sorting column failover. These are all `DataObjects` and, as such, all will contain this column.
- Still add a new config option, but instead of `default_sort_column` (which is somewhat ambiguous in meaning), we use the name such as: `sort_deterministic`, `sort_secondary_id` or  `sort_failover` (or similar).
  - Note: I'd stay away from the `default_sort_` prefix since we're not defaulting anything per se (as you'll see below). Instead, we're enabling/disabling a deterministic sorting algorithm at the database level, which (yes) does mean we have to inject extra fluff into the `ORDER BY` clause and (yes) this setting is enabled by default. However, this is a boolean (noted below)).
- Per above (and discussion): This should be an on/off setting and (per above) this should default to `true`.
- **Direction:** We can automatically adjust the direction of secondary `ID` sort failover by defaulting to `ASC` in our algorithm and then parsing the existing order setting in our query prior to final execution for a direction and, if not blank/ASC, we can then set it to `DESC`.
- **Disabling/overriding:**  This can be done one of either one or two ways (uncertain on second):
  - Override the `sort_deterministic` setting and set it to `false`.
  - Parse final order clause for any secondary sort fields and, if they exist, remove.
    - **NOTE:** I'm unsure if this is necessary, as you can have as many additional sort columns as you like (as far as I'm aware) and a developer with the confidence of adding a secondary sort column may already have a very unique sort, so adding `ID` may be redundant for several reasons but not necessarily negative or have an impact on performance.
    - At minimum, if we retain `ID` even if a secondary column is manually provided, we may still want to do a quick de-dupe check to ensure this `ID` column isn't already specified (to prevent accidentally overriding a contrary direction, e.g. `DESC` instead of `ASC`).

Beyond that, I think we'd be set. Thoughts?

**Comment by flamerohr** [comment]
I agree with @patricknelson that having another config setting called `default_sort_something` is ambiguous and conflicts with the existing `default_sort`, but the idea is the same - to have a failover column to enable deterministic sorting.
Having `ID` as a default for whatever this setting is called is a good idea, as all DataObjects have this (maybe rare exceptions).

If the idea can be kept simple such as a single config `private static $determined_sort = ['ID' => 'ASC'];` and that can be extended/disabled as desired by developers through php or yml that would be good - yes you can set a config to `false` or `null` in yml in ss4.

**Comment by sunnysideup** [comment]
I think some research may be useful:

* what are the consequences of adding a sort by ID ASC to all sql statements (after any other sort statements - if any), so that you always have a predictable outcome.  This would be my solution and apart from performance reductions, I can't see any other reason why this is not a good solution.

* I noticed that dataobjects do sort randomly but pages do not and I wonder why that is (maybe to do with joins?)

Also - while no one has noticed this issues, this is not a benign one.  At least two of my clients noticed it (and were unable to edit data because of it) in gridfields (where I had set the sort order to a field that was something like a date field).

**Comment by sunnysideup** [comment]
Adding an index to SiteTree.Sort should be a no-brainer AFAIK.

**Comment by patricknelson** [comment]
Actually, what if we just had a `default_sort_secondary` field? This could not only be consistent with the existing `default_sort` field, but can be automatically defaulted to `ID ASC`. Let me explain some scenarios and why I think this might work.

**Functionality:**

- Default value for `default_sort_secondary` is  `ID ASC`.
- Behind the scenes, the secondary column and direction are parsed separately before being handed to the database. This way, we can handle scenarios when the user begins to customize sorting. The sort direction is just an initial value that we start with and then _negate_ when the user negates the initial direction. That way if the user ever selects another primary sort column or ever chooses to negate the direction, we just negate our initial direction. No magic/mystery to this, just keep it logical. So, we're just shifting perspective. Logically, if you reverse the current sort, the secondary sort must also be reversed as well for consistency.
- If `default_sort` is defined with a secondary sort value, that secondary sort will be used. e.g. `Date ASC, Title ASC` will imply a `default_sort_secondary` value of `Title ASC`.
- Conflict resolution: If current sort column (e.g. `Title`) is equal to the above resolved secondary sort, secondary sort fails over to the `ID` column (since we know `ID` will always be unique).

This satisfies:

- Nomenclature consistency (`default_sort_secondary` as an extension to `default_sort`).
- Sane initial defaults which are less magical (e.g. if the user sets up a secondary sort column in `default_sort`, this is parsed/retained in `default_sort_secondary` and functionality is extended).
- Logical, consistent and deterministic results for pagination.

What do you guys think about that? Am I missing any scenarios or edge cases, possibly?

**Comment by sunnysideup** [comment]
Any movement ?

**Comment by maxime-rainville** [comment]
Just skim read through this delightful thread. While the suggested feature would provide some benefit for some very specific use cases, it sounds like it's somewhat low value for most common scenarios. Those people who would benefit from this have a relatively easy alternative (just manually include a ID sort in their queries).

My instinct would be to close this issue has a "won't fix" since there's no clear agreed upon action to take.

**Comment by patricknelson** [comment]
Reread this. That would be fine with me as it's still possible (as the developer) to manually define the secondary sort column via `::$default_sort`. However, how do you (as SilverStripe) conclude this and resolve @unclecheese's point that `DataList` is intended to be deterministic, per https://github.com/silverstripe/silverstripe-framework/issues/7249#issuecomment-321949846? 

That is ultimately what I think this issue morphed into, which makes it more broad and less about a specific use case like you'd find just in `GridField` pagination (or whatever). It's probably still low priority since while it may occur frequently, it's easy to fix (i.e. via best practices @tractorcow mentioned), but would still represent an important low level change, considering the breadth of it. 

So to summarize: Is `DataList` deterministic, despite the underlying database layer or not?

**Comment by jonom** [comment]
FWIW I experienced this bug on SS4 today with 'out of the box' models - in the Security->Users section. A site had many members with no first or last name, so I got repeat/missing members on subsequent pages. Looks like it would [remain an issue on SS5](https://github.com/silverstripe/silverstripe-framework/blob/5/src/Security/Member.php#L96).

Putting aside the bigger debate for now, quick win would be to fix up all core models to ensure consistent sorting (e.g. add ID and maybe Email to Member default_sort), and maybe update the docs to flag this issue as suggested previously?

Edit: I guess that would only help for literally the 'default sort' case. So if I applied a sort on first/last name in my example by clicking a column header, the issue would come back because ID would no longer be included in the sort clause.

**Comment by jonom** [comment]
Think the scope is bigger than the title suggests as it's not just GridField (it's all DataLists) and not just about default_sort. Even if you fix your default_sort to include ID, that won't come for the ride if you choose a different sort by using the sortable headers that are baked in to Silverstripe CMS.

Another pain point is that it [doesn't appear](https://docs.silverstripe.org/en/5/developer_guides/graphql/working_with_dataobjects/query_plugins/#the-sort-plugin) that you can sort on multiple fields with the GraphQL module. `sort` expects an object instead of an array so passing multiple fields already wouldn't let you specify which field should come first, but in any case if you do it seems that only the field name that's alphabetically last is used for the sort. So overall this doesn't appear to be an issue that developers can easily workaround by tweaking a few sort clauses in their own code. _Edit: this looks at least partially to [be a bug](https://github.com/silverstripe/silverstripe-graphql/pull/541)._

**Comment by sunnysideup** [comment]
BUMP BUMP BUMP
